<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="This project contains data and source code generated by Dr.
    Su Wu, Ph.D., as part of her academic research work at Harvard Medical
    School.">
<title>A R package to analyze eIF4F subunits in human cancers • eIF4F.analysis</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A R package to analyze eIF4F subunits in human cancers">
<meta property="og:description" content="This project contains data and source code generated by Dr.
    Su Wu, Ph.D., as part of her academic research work at Harvard Medical
    School.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">eIF4F.analysis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/eIF4F_analysis.html">eIF4F_analysis</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/a3609640/eIF4F.analysis/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="eif4fanalysis">eIF4F.analysis<a class="anchor" aria-label="anchor" href="#eif4fanalysis"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>The goal of eIF4F.analysis is to …</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of eIF4F.analysis from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"a3609640/eIF4F.analysis"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="before-you-begin">Before you begin<a class="anchor" aria-label="anchor" href="#before-you-begin"></a>
</h2>
<p>Perform the following steps to ensure the proper operation of this package.</p>
<p><a href="#system-requirements">System requirements</a></p>
<p><a href="#install-rstudior">Install RStudio/R</a></p>
<p><a href="#download-data-sets">Download data-sets</a></p>
<p><a href="#file-directories">File directories</a></p>
<p><a href="#install-dependent-libraries">Install dependent libraries</a></p>
<p><a href="#load-package">Load package</a></p>
<p><a href="#session-information">Session information</a></p>
</div>
<div class="section level2">
<h2 id="system-requirements">System requirements<a class="anchor" aria-label="anchor" href="#system-requirements"></a>
</h2>
<p>This project makes use of various resource-intensive R packages, which carries relatively high demands for compute, RAM, and disk I/O resources (but not for graphics resources). Nonetheless, the necessary hardware is attainable in high-end consumer-grade systems.</p>
<div class="section level3">
<h3 id="description-of-development-systems">Description of development systems<a class="anchor" aria-label="anchor" href="#description-of-development-systems"></a>
</h3>
<p>The following systems have been used to execute the R scripts in this project:</p>
<ol style="list-style-type: decimal">
<li>
<p>(verified) System76 “Serval” mobile workstation</p>
<ul>
<li>Intel i7-8700k CPU</li>
<li>64GB RAM (DDR4-3000, non-ECC)</li>
<li>Samsung NVMe Pro SSD</li>
<li>Pop!_OS 20.04 LTS</li>
<li>RStudio</li>
<li>R 4.1.1</li>
</ul>
</li>
<li>
<p>(verified) PowerSpec G460 desktop computer</p>
<ul>
<li>Intel i7-8700k CPU</li>
<li>64GB RAM (DDR4-3200, non-ECC)</li>
<li>Intel M.2 SATA SSD</li>
<li>Samsung NVMe Evo+ SSD</li>
<li>Windows 10 Pro</li>
<li>RStudio for Windows</li>
<li>R 4.0.3</li>
</ul>
</li>
</ol>
<p>Additional details of these environments are provided in the “Session Information” section below.</p>
</div>
</div>
<div class="section level2">
<h2 id="install-rstudior">Install RStudio/R<a class="anchor" aria-label="anchor" href="#install-rstudior"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Download &amp; install R 4.1, if not already installed.</li>
<li>Download &amp; install RStudio, if not already installed. <a href="https://www.rstudio.com/products/rstudio/download/" class="external-link uri">https://www.rstudio.com/products/rstudio/download/</a>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="download-data-sets">Download data-sets<a class="anchor" aria-label="anchor" href="#download-data-sets"></a>
</h2>
<p>Run <code>Download.sh</code> from <code>Script</code> folder of our GitHub repository.</p>
<p>Download.sh</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> ~/github/eIF4F.analysis/Script/Download.sh</span></code></pre></div>
<p><code>Download.sh</code> is a bash script to download all needed datasets (TCGA, GTEx, CPTAC, CCLE and etc) from URLs and unzip them. <code>Download.sh</code> will create the <code>~/Downloads/EIF_data</code> directory to store all downloaded datasets.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/sh</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">## download all datasets from the following weblinks</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co">### create the directory to store all downloaded datasets</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="bu">readonly</span> <span class="va">DATA_FILE_DIRECTORY</span><span class="op">=</span><span class="st">"</span><span class="va">${HOME}</span><span class="st">/Downloads/EIF_data"</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co">### TCGA and GTEX DATA</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="co">#### TCGA CNV dataset (thresholded)</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://tcga.xenahubs.net/download/TCGA.PANCAN.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes.gz <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="co">#### TCGA CNV dataset</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://tcga.xenahubs.net/download/TCGA.PANCAN.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="co">#### TCGA CNV ratio dataset</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://pancanatlas.xenahubs.net/download/broad.mit.edu_PANCAN_Genome_Wide_SNP_6_whitelisted.gene.xena.gz <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="co">#### TCGA RNA-Seq dataset</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://tcga-pancan-atlas-hub.s3.us-east-1.amazonaws.com/download/EB%2B%2BAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena.gz <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="co">#### TCGA sample type annotation</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://pancanatlas.xenahubs.net/download/TCGA_phenotype_denseDataOnlyDownload.tsv.gz <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="co">#### TCGA OS data ##</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://tcga-pancan-atlas-hub.s3.us-east-1.amazonaws.com/download/Survival_SupplementalTable_S1_20171025_xena_sp <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="co">#### TCGA and GTEX RNA-Seq dataset</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://toil.xenahubs.net/download/TcgaTargetGtex_RSEM_Hugo_norm_count.gz <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a><span class="co">#### TCGA and GTEX sample type annotation</span></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://toil.xenahubs.net/download/TcgaTargetGTEX_phenotype.txt.gz <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a><span class="co">### CPTAC DATA</span></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a><span class="co">#### CPTAC LUAD RNA-Seq data (Gillette et al., 2020)</span></span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://github.com/a3609640/EIF-analysis/raw/master/LUAD%20Data/RNA.xlsx <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a><span class="co">#### CPTAC LUAD Proteomics (Gillette et al., 2020)</span></span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://github.com/a3609640/EIF-analysis/raw/master/LUAD%20Data/Protein.xlsx <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a><span class="co">#### CPTAC LUAD Proteomics</span></span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://cptc-xfer.uis.georgetown.edu/publicData/Phase_III_Data/CPTAC_LUAD_S046/CPTAC_LUAD_Proteome_CDAP_Protein_Report.r1/CPTAC3_Lung_Adeno_Carcinoma_Proteome.tmt10.tsv <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a><span class="co">#### CPTAC LUAD Phosproteomics (Gillette et al., 2020)</span></span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://github.com/a3609640/EIF-analysis/raw/master/LUAD%20Data/Phos.xlsx <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a><span class="co">#### CPTAC LUAD Sample Annotation</span></span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://cptc-xfer.uis.georgetown.edu/publicData/Phase_III_Data/CPTAC_LUAD_S046/CPTAC_LUAD_metadata/S046_BI_CPTAC3_LUAD_Discovery_Cohort_Samples_r1_May2019.xlsx <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a><span class="co">#### CPTAC Clinical Data</span></span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://cptc-xfer.uis.georgetown.edu/publicData/Phase_III_Data/CPTAC_LUAD_S046/CPTAC_LUAD_metadata/S046_BI_CPTAC3_LUAD_Discovery_Cohort_Clinical_Data_r1_May2019.xlsx <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a><span class="co">### CCLE DATA</span></span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a><span class="co">#### CCLE RNA-Seq data from DepMap Public 20Q4 20Q3</span></span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a><span class="co">#wget -O CCLE_expression_full.csv https://ndownloader.figshare.com/files/#27902097 -P "${DATA_FILE_DIRECTORY}" #DepMap Public 21Q2</span></span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://ndownloader.figshare.com/files/24613349 <span class="at">-O</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">/CCLE_expression_full.csv"</span> <span class="co">#DepMap Public 20Q3</span></span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a><span class="co">#### CCLE annotation data</span></span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a><span class="co">#wget -O sample_info.csv https://ndownloader.figshare.com/files/27902376 -P #"${DATA_FILE_DIRECTORY}" #DepMap Public 21Q2</span></span>
<span id="cb3-65"><a href="#cb3-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-66"><a href="#cb3-66" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://ndownloader.figshare.com/files/24613394 <span class="at">-O</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">/sample_info.csv"</span> <span class="co">#DepMap Public 20Q3</span></span>
<span id="cb3-67"><a href="#cb3-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-68"><a href="#cb3-68" aria-hidden="true" tabindex="-1"></a><span class="co">#### CCLE proteomics data</span></span>
<span id="cb3-69"><a href="#cb3-69" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://gygi.hms.harvard.edu/data/ccle/protein_quant_current_normalized.csv.gz <span class="at">-P</span> <span class="st">"</span><span class="va">${DATA_FILE_DIRECTORY}</span><span class="st">"</span></span>
<span id="cb3-70"><a href="#cb3-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-71"><a href="#cb3-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-72"><a href="#cb3-72" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> <span class="va">${DATA_FILE_DIRECTORY}</span>/<span class="pp">*</span>.gz</span></code></pre></div>
<p><strong>CRITICAL</strong>: If the root directory path <code>~/Downloads/EIF_data</code> does not suit, they may be adjusted trivially in these lines near the top of the <code>Download.sh</code> script.</p>
</div>
<div class="section level2">
<h2 id="file-directories">File directories<a class="anchor" aria-label="anchor" href="#file-directories"></a>
</h2>
<p>Confirm the directories for input and output files. The directories for input and output files are defined in the <code>Load.R</code> as following.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>data_file_directory <span class="ot">&lt;-</span> <span class="st">"~/Downloads/EIF_data"</span> </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>output_directory <span class="ot">&lt;-</span> <span class="st">"~/Documents/EIF_output"</span></span></code></pre></div>
<p><strong>CRITICAL</strong>: If the root directory paths <code>~/Download/EIF_data</code> and <code>~/Documents/EIF_output</code> do not suit, they may be adjusted trivially in these lines near the top of the <code>Download.sh</code> and <code>Load.R</code> scripts.</p>
</div>
<div class="section level2">
<h2 id="install-dependent-libraries">Install dependent libraries<a class="anchor" aria-label="anchor" href="#install-dependent-libraries"></a>
</h2>
<p>The work here depends upon many R libraries. The following command may be a useful way to install them all.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># use Bioconductor version 3.14 for package installation</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.14"</span><span class="op">)</span>

<span class="va">bio_pkgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
  <span class="st">"AnnotationDbi"</span>, <span class="st">"circlize"</span>,  <span class="st">"clusterProfiler"</span>, <span class="st">"ComplexHeatmap"</span>, <span class="st">"corrplot"</span>, 
  <span class="st">"data.table"</span>, <span class="st">"dplyr"</span>, <span class="st">"EnvStats"</span>, <span class="st">"eulerr"</span>, <span class="st">"factoextra"</span>, <span class="st">"FactoMineR"</span>, 
  <span class="st">"forcats"</span>, <span class="st">"forestplot"</span>, <span class="st">"ggfortify"</span>, <span class="st">"ggplot2"</span>, <span class="st">"ggpubr"</span>,  <span class="st">"graphics"</span>, 
  <span class="st">"grDevices"</span>, <span class="st">"grid"</span>, <span class="st">"limma"</span>, <span class="st">"missMDA"</span>, <span class="st">"org.Hs.eg.db"</span>, <span class="st">"purrr"</span>, 
  <span class="st">"RColorBrewer"</span>, <span class="st">"ReactomePA"</span>, <span class="st">"readr"</span>,  <span class="st">"readxl"</span>, <span class="st">"reshape2"</span>, <span class="st">"scales"</span>, 
  <span class="st">"stats"</span>, <span class="st">"stringr"</span>, <span class="st">"survival"</span>, <span class="st">"survivalAnalysis"</span>, <span class="st">"tibble"</span>, <span class="st">"tidyr"</span>, 
  <span class="st">"tidyselect"</span><span class="op">)</span>

<span class="co"># install required packages</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="va">bio_pkgs</span><span class="op">)</span>

<span class="co"># load required packages</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bio_pkgs</span>, <span class="va">require</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-package">Load package<a class="anchor" aria-label="anchor" href="#load-package"></a>
</h2>
<p>Load <code>eIF4F.analysis</code> in the R console.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/a3609640/eIF4F.analysis" class="external-link">eIF4F.analysis</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>The version information of R, Linux and attached or loaded packages for developing this package is the following.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">sessionInfo</span>(<span class="at">package =</span> <span class="st">"eIF4F.analysis"</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>R version <span class="dv">4</span>.<span class="fl">1.1</span> (<span class="dv">2021-08-10</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>Platform<span class="sc">:</span> x86_64<span class="sc">-</span>pc<span class="sc">-</span>linux<span class="sc">-</span><span class="fu">gnu</span> (<span class="dv">64</span><span class="sc">-</span>bit)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>Running under<span class="sc">:</span> Pop<span class="sc">!</span>_OS <span class="fl">20.04</span> LTS</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>Matrix products<span class="sc">:</span> default</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>BLAS<span class="sc">:</span>   <span class="er">/</span>usr<span class="sc">/</span>lib<span class="sc">/</span>x86_64<span class="sc">-</span>linux<span class="sc">-</span>gnu<span class="sc">/</span>blas<span class="sc">/</span>libblas.so.<span class="dv">3</span>.<span class="fl">9.0</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>LAPACK<span class="sc">:</span> <span class="er">/</span>usr<span class="sc">/</span>lib<span class="sc">/</span>x86_64<span class="sc">-</span>linux<span class="sc">-</span>gnu<span class="sc">/</span>lapack<span class="sc">/</span>liblapack.so.<span class="dv">3</span>.<span class="fl">9.0</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>locale<span class="sc">:</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a> [<span class="dv">1</span>] LC_CTYPE<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>       LC_NUMERIC<span class="ot">=</span>C               LC_TIME<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>        LC_COLLATE<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>    </span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a> [<span class="dv">5</span>] LC_MONETARY<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>    LC_MESSAGES<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>    LC_PAPER<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>       LC_NAME<span class="ot">=</span>C                 </span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a> [<span class="dv">9</span>] LC_ADDRESS<span class="ot">=</span>C               LC_TELEPHONE<span class="ot">=</span>C             LC_MEASUREMENT<span class="ot">=</span>en_US.UTF<span class="dv">-8</span> LC_IDENTIFICATION<span class="ot">=</span>C       </span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>attached base packages<span class="sc">:</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="fu">character</span>(<span class="dv">0</span>)</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>other attached packages<span class="sc">:</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] eIF4F.analysis_0.<span class="fl">1.0</span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">1</span>] utf8_1.<span class="fl">2.2</span>             tidyselect_1.<span class="fl">1.1</span>       RSQLite_2.<span class="fl">2.9</span>          AnnotationDbi_1.<span class="fl">56.2</span>   htmlwidgets_1.<span class="fl">5.4</span>     </span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>  [<span class="dv">6</span>] FactoMineR_2<span class="fl">.4</span>         grid_4.<span class="fl">1.1</span>             BiocParallel_1.<span class="fl">28.3</span>    scatterpie_0.<span class="fl">1.7</span>       munsell_0.<span class="fl">5.0</span>         </span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a> [<span class="dv">11</span>] codetools_0<span class="fl">.2</span><span class="dv">-18</span>       DT_0<span class="fl">.20</span>                withr_2.<span class="fl">4.3</span>            colorspace_2<span class="fl">.0</span><span class="dv">-2</span>       GOSemSim_2.<span class="fl">20.0</span>       </span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a> [<span class="dv">16</span>] Biobase_2.<span class="fl">54.0</span>         ggfortify_0.<span class="fl">4.13</span>       knitr_1<span class="fl">.37</span>             rstudioapi_0<span class="fl">.13</span>        leaps_3<span class="fl">.1</span>             </span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a> [<span class="dv">21</span>] stats4_4.<span class="fl">1.1</span>           ggsignif_0.<span class="fl">6.3</span>         DOSE_3.<span class="fl">20.1</span>            GenomeInfoDbData_1.<span class="fl">2.7</span> KMsurv_0<span class="fl">.1</span><span class="dv">-5</span>          </span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a> [<span class="dv">26</span>] polyclip_1<span class="fl">.10</span><span class="dv">-0</span>        stats_4.<span class="fl">1.1</span>            bit64_4.<span class="fl">0.5</span>            farver_2.<span class="fl">1.0</span>           rprojroot_2.<span class="fl">0.2</span>       </span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a> [<span class="dv">31</span>] downloader_0<span class="fl">.4</span>         survivalAnalysis_0.<span class="fl">2.0</span> vctrs_0.<span class="fl">3.8</span>            treeio_1.<span class="fl">18.1</span>          generics_0.<span class="fl">1.1</span>        </span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a> [<span class="dv">36</span>] xfun_0<span class="fl">.29</span>              R6_2.<span class="fl">5.1</span>               doParallel_1.<span class="fl">0.16</span>      GenomeInfoDb_1.<span class="fl">30.0</span>    clue_0<span class="fl">.3</span><span class="dv">-60</span>           </span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a> [<span class="dv">41</span>] graphlayouts_0.<span class="fl">7.2</span>     bitops_1<span class="fl">.0</span><span class="dv">-7</span>           cachem_1.<span class="fl">0.6</span>           fgsea_1.<span class="fl">20.0</span>           gridGraphics_0<span class="fl">.5</span><span class="dv">-1</span>    </span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a> [<span class="dv">46</span>] assertthat_0.<span class="fl">2.1</span>       scales_1.<span class="fl">1.1</span>           ggraph_2.<span class="fl">0.5</span>           enrichplot_1.<span class="fl">14.1</span>      gtable_0.<span class="fl">3.0</span>          </span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a> [<span class="dv">51</span>] tidygraph_1.<span class="fl">2.0</span>        rlang_0.<span class="fl">4.12</span>           utils_4.<span class="fl">1.1</span>            methods_4.<span class="fl">1.1</span>          scatterplot3d_0<span class="fl">.3</span><span class="dv">-41</span>  </span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a> [<span class="dv">56</span>] eulerr_6.<span class="fl">1.1</span>           GlobalOptions_0.<span class="fl">1.2</span>    splines_4.<span class="fl">1.1</span>          rstatix_0.<span class="fl">7.0</span>          lazyeval_0.<span class="fl">2.2</span>        </span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a> [<span class="dv">61</span>] broom_0.<span class="fl">7.10</span>           checkmate_2.<span class="fl">0.0</span>        reshape2_1.<span class="fl">4.4</span>         abind_1<span class="fl">.4</span><span class="dv">-5</span>            backports_1.<span class="fl">4.1</span>       </span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a> [<span class="dv">66</span>] qvalue_2.<span class="fl">26.0</span>          base_4.<span class="fl">1.1</span>             clusterProfiler_4.<span class="fl">2.1</span>  tools_4.<span class="fl">1.1</span>            tidytidbits_0.<span class="fl">2.3</span>     </span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a> [<span class="dv">71</span>] ggplotify_0.<span class="fl">1.0</span>        ggplot2_3.<span class="fl">3.5</span>          ellipsis_0.<span class="fl">3.2</span>         RColorBrewer_1<span class="fl">.1</span><span class="dv">-2</span>     BiocGenerics_0.<span class="fl">40.0</span>   </span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a> [<span class="dv">76</span>] Rcpp_1.<span class="fl">0.7</span>             plyr_1.<span class="fl">8.6</span>             zlibbioc_1.<span class="fl">40.0</span>        purrr_0.<span class="fl">3.4</span>            RCurl_1<span class="fl">.98-1.5</span>        </span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a> [<span class="dv">81</span>] ggpubr_0.<span class="fl">4.0</span>           GetoptLong_1.<span class="fl">0.5</span>       viridis_0.<span class="fl">6.2</span>          cowplot_1.<span class="fl">1.1</span>          zoo_1<span class="fl">.8</span><span class="dv">-9</span>             </span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a> [<span class="dv">86</span>] S4Vectors_0.<span class="fl">32.3</span>       ggrepel_0.<span class="fl">9.1</span>          cluster_2.<span class="fl">1.2</span>          factoextra_1.<span class="fl">0.7</span>       magrittr_2.<span class="fl">0.1</span>        </span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a> [<span class="dv">91</span>] data.table_1.<span class="fl">14.2</span>      DO.db_2<span class="fl">.9</span>              forestplot_2.<span class="fl">0.1</span>       circlize_0.<span class="fl">4.13</span>        reactome.db_1.<span class="fl">77.0</span>    </span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a> [<span class="dv">96</span>] survminer_0.<span class="fl">4.9</span>        mvtnorm_1<span class="fl">.1</span><span class="dv">-3</span>          matrixStats_0.<span class="fl">61.0</span>     pkgload_1.<span class="fl">2.4</span>          xtable_1<span class="fl">.8</span><span class="dv">-4</span>          </span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a>[<span class="dv">101</span>] hms_1.<span class="fl">1.1</span>              patchwork_1.<span class="fl">1.1</span>        readxl_1.<span class="fl">3.1</span>           IRanges_2.<span class="fl">28.0</span>         gridExtra_2<span class="fl">.3</span>         </span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a>[<span class="dv">106</span>] missMDA_1<span class="fl">.18</span>           shape_1.<span class="fl">4.6</span>            testthat_3.<span class="fl">1.1</span>         compiler_4.<span class="fl">1.1</span>         tibble_3.<span class="fl">1.6</span>          </span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>[<span class="dv">111</span>] mice_3.<span class="fl">14.0</span>            crayon_1.<span class="fl">4.2</span>           shadowtext_0.<span class="fl">1.0</span>       htmltools_0.<span class="fl">5.2</span>        ggfun_0.<span class="fl">0.4</span>           </span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>[<span class="dv">116</span>] tzdb_0.<span class="fl">2.0</span>             tidyr_1.<span class="fl">1.4</span>            aplot_0.<span class="fl">1.1</span>            ReactomePA_1.<span class="fl">38.0</span>      DBI_1.<span class="fl">1.2</span>             </span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>[<span class="dv">121</span>] tweenr_1.<span class="fl">0.2</span>           corrplot_0<span class="fl">.92</span>          EnvStats_2.<span class="fl">4.0</span>         ComplexHeatmap_2.<span class="fl">10.0</span>  MASS_7<span class="fl">.3</span><span class="dv">-54</span>           </span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>[<span class="dv">126</span>] rappdirs_0.<span class="fl">3.3</span>         Matrix_1<span class="fl">.4</span><span class="dv">-0</span>           car_3<span class="fl">.0</span><span class="dv">-12</span>             readr_2.<span class="fl">1.1</span>            cli_3.<span class="fl">1.0</span>             </span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>[<span class="dv">131</span>] datasets_4.<span class="fl">1.1</span>         grDevices_4.<span class="fl">1.1</span>        parallel_4.<span class="fl">1.1</span>         igraph_1.<span class="fl">2.10</span>          km.ci_0<span class="fl">.5</span><span class="dv">-2</span>           </span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a>[<span class="dv">136</span>] forcats_0.<span class="fl">5.1</span>          pkgconfig_2.<span class="fl">0.3</span>        flashClust_1<span class="fl">.01</span><span class="dv">-2</span>      foreach_1.<span class="fl">5.1</span>          ggtree_3.<span class="fl">2.1</span>          </span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>[<span class="dv">141</span>] XVector_0.<span class="fl">34.0</span>         graphics_4.<span class="fl">1.1</span>         yulab.utils_0.<span class="fl">0.4</span>      stringr_1.<span class="fl">4.0</span>          digest_0.<span class="fl">6.29</span>         </span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a>[<span class="dv">146</span>] graph_1.<span class="fl">72.0</span>           Biostrings_2.<span class="fl">62.0</span>      cellranger_1.<span class="fl">1.0</span>       fastmatch_1<span class="fl">.1</span><span class="dv">-3</span>        survMisc_0.<span class="fl">5.5</span>        </span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a>[<span class="dv">151</span>] tidytree_0.<span class="fl">3.6</span>         graphite_1.<span class="fl">40.0</span>        rjson_0.<span class="fl">2.20</span>           lifecycle_1.<span class="fl">0.1</span>        nlme_3<span class="fl">.1</span><span class="dv">-153</span>          </span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a>[<span class="dv">156</span>] jsonlite_1.<span class="fl">7.2</span>         carData_3<span class="fl">.0</span><span class="dv">-4</span>          desc_1.<span class="fl">4.0</span>             viridisLite_0.<span class="fl">4.0</span>      limma_3.<span class="fl">50.0</span>          </span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a>[<span class="dv">161</span>] fansi_0.<span class="fl">5.0</span>            pillar_1.<span class="fl">6.4</span>           lattice_0<span class="fl">.20</span><span class="dv">-45</span>        KEGGREST_1.<span class="fl">34.0</span>        fastmap_1.<span class="fl">1.0</span>         </span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a>[<span class="dv">166</span>] httr_1.<span class="fl">4.2</span>             survival_3<span class="fl">.2</span><span class="dv">-13</span>        GO.db_3.<span class="fl">14.0</span>           glue_1.<span class="fl">6.0</span>             png_0<span class="fl">.1</span><span class="dv">-7</span>             </span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a>[<span class="dv">171</span>] iterators_1.<span class="fl">0.13</span>       bit_4.<span class="fl">0.4</span>              ggforce_0.<span class="fl">3.3</span>          stringi_1.<span class="fl">7.6</span>          blob_1.<span class="fl">2.2</span>            </span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a>[<span class="dv">176</span>] org.Hs.eg.db_3.<span class="fl">14.0</span>    memoise_2.<span class="fl">0.1</span>          dplyr_1.<span class="fl">0.7</span>            ape_5<span class="fl">.6</span>   </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorials">
<strong>Tutorials</strong><a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h2>
<p>Open the <code>Analysis.R</code> from <code>Script</code> folder from our GitHub repository. This script contains the command lines to execute all analyses.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/a3609640/eIF4F.analysis/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/a3609640/eIF4F.analysis/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>Apache License (&gt;= 2)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing eIF4F.analysis</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Su Wu <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Su Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
