# PCA on EIF4F expression in GTEx, TCGA, and combined studies ------------------

# This R script contains four sections.
# (1) select TCGA RNA-seq dataset for PCA.
# (2) perform PCA or imputed PCA and plot the PCA results as biplots or selected biplots
# (3) master functions to execute a pipeline of functions to select related RNAseq data
# for PCA and plot with supply of EIF4F gene names as values of the arguments.
# (4) wrapper function to call all master functions with inputs


## Select TCGA RNA-seq data for PCA ============================================

#' Select RNAseq data based on the sample types
#' @description This function selects the RNAseq data from different sample types:
#' primary tumors (TCGA), metastatic tumors (TCGA), healthy tissues (GTEX)
#' or all sample types combined (all), from dataframe \code{.TCGA_GTEX_sampletype_subset}.
#'
#' It should not be used directly, only inside \code{\link{.plot_PCA_TCGA_GTEX}} or
#' \code{\link{.plot_PCA_TCGA_GTEX_tumor}} function.
#' @param df \code{.TCGA_GTEX_sampletype_subset} generated inside \code{\link{.plot_PCA_TCGA_GTEX}}
#' @param x selected sample types
#' @return a subset data frame from .TCGA_GTEX_sampletype_subset for indicate sample types
#' @examples \dontrun{
#' .get_df_subset(.TCGA_GTEX_sampletype_subset, "Primary Tumor (TCGA)")
#' }
#' @keywords internal
#'
.get_df_subset <- function(df, x) {
  df1 <- df %>%
    dplyr::filter(if (x != "All") .data$sample.type == x else TRUE) %>%
    droplevels()
  df2 <- df1 %>% dplyr::select_if(is.numeric)
  output <- list(df2, df1)
  return(output)
}

## Perform PCA or imputed PCA and plot the PCA results =========================

#' Perform PCA on RNAseq data
#' @description This function perform PCA  on data generated by \code{\link{.get_df_subset}}.
#' It should not be used directly, only inside \code{\link{.plot_PCA_TCGA_GTEX}} or
#' \code{\link{.plot_PCA_TCGA_GTEX_tumor}} function.
#' @param df \code{df} generated inside \code{\link{.get_df_subset}}
#' @param x number of components kept in the final results
#' @return PCA results
#' @importFrom FactoMineR PCA
#' @examples \dontrun{
#' .RNAseq_PCA(df[[1]], 10)
#' }
#' @keywords internal
#'
.RNAseq_PCA <- function(df, x) {
  res.pca <- FactoMineR::PCA(df, # remove column with characters
    scale.unit = TRUE,
    ncp = x,
    graph = FALSE
  )
  return(res.pca)
}

#' Perform imputPCA on proteomics data
#' @description This function perform impute PCA on data generated by \code{\link{.get_df_subset}}.
#' It should not be used directly, only inside \code{\link{.plot_PCA_CPTAC_LUAD}} function.
#' @param df \code{CPTAC.LUAD.Proteomics.Sample.subset} generated inside \code{\link{.plot_PCA_CPTAC_LUAD}}
#' @param x maximum number of components kept in the final results
#' @return imputed PCA results
#' @importFrom missMDA estim_ncpPCA imputePCA
#' @examples \dontrun{
#' .protein_imputePCA(CPTAC.LUAD.Proteomics.Sample.subset)
#' }
#' @keywords internal
#'
.protein_imputePCA <- function(df, x) {
  # Impute the missing values of a dataset with the Principal Components Analysis model
  nb <- missMDA::estim_ncpPCA(
    df %>% dplyr::select_if(is.numeric),
    # CPTAC.LUAD.Proteomics.Sample[1:(length(CPTAC.LUAD.Proteomics.Sample) - 1)],
    ncp.max = x
  ) # estimate the number of dimensions to impute
  res.comp <- missMDA::imputePCA(
    df %>% dplyr::select_if(is.numeric),
    # CPTAC.LUAD.Proteomics.Sample[1:(length(CPTAC.LUAD.Proteomics.Sample) - 1)],
    ncp = nb$ncp
  )
  res.pca <- PCA(res.comp$completeObs,
    scale.unit = TRUE,
    ncp = x,
    graph = FALSE
  )
  return(res.pca)
}

#' Plot PCA results as biplot, scree and matrix plots
#' @description This function draw biplot, screen and matrix plots for PCA results
#' from selected RNAseq data generated from \code{\link{.get_df_subset}}.
#'
#' @param res.pca PCA results from selected RNAseq data
#' @param df selected RNAseq data with sample type annotation
#' @param x selected sample types for plot label
#' @param y type of primary diseases or healthy tissues for color of individuals
#' @param color color scheme used for individual sample on the PCA
#' @param folder sub directory name to store the output files
#' @return PCA bioplot, matrix, scree plot
#' @importFrom corrplot corrplot
#' @importFrom factoextra fviz_contrib fviz_eig fviz_pca_biplot get_pca_var
#' @examples \dontrun{
#' .biplot(
#'   res.pca = .RNAseq_PCA(df[[1]], 10), df = df[[2]],
#'   x = "Metastatic Tumor (TCGA)", y = "primary.disease", color = col_vector, folder = "TCGA"
#' )
#' }
#' @keywords internal
#'
.biplot <- function(res.pca, df, x, y, color, folder) {
  biplot <- fviz_pca_biplot(res.pca,
    axes = c(1, 2),
    labelsize = 5,
    col.ind = df[, y],
    # title = paste("PCA - Biplot (", x, ")"),
    palette = color,
    pointshape = 20,
    pointsize = 0.75,
    label = "var",
    col.var = "black",
    repel = TRUE
  ) +
    {
      if (x == "All") {
        labs(title = "PCA - Biplot (Healthy Tissues + Tumors)")
      } else {
        labs(title = paste0("PCA - Biplot (", x, ")"))
      }
    } +
    # xlim(-7, 8) +
    # ylim(-6, 7.5) + # for EIF 8
    theme_classic() +
    theme(
      plot.background = element_blank(),
      plot.title = black_bold_16,
      panel.background = element_rect(
        fill = "transparent",
        color = "black",
        size = 1
      ),
      axis.title.x = black_bold_16,
      axis.title.y = black_bold_16,
      axis.text.x = black_bold_16,
      axis.text.y = black_bold_16,
      legend.title = element_blank(),
      legend.position = c(0, 0),
      legend.justification = c(0, 0),
      legend.background = element_blank(),
      legend.text = black_bold_16
    )
  print(biplot)
  ggplot2::ggsave(
    path = file.path(output.directory, "PCA", folder),
    filename = paste("PCA", x, ".pdf"),
    plot = biplot,
    width = 8,
    height = 8,
    useDingbats = FALSE
  )

  eig <- fviz_eig(res.pca,
    labelsize = 6,
    geom = "bar",
    width = 0.7,
    addlabels = TRUE
  ) +
    {
      if (x == "All") {
        labs(title = "Scree plot (Healthy Tissues + Tumors)")
      } else {
        labs(title = paste0("Scree plot (", x, ")"))
      }
    } +
    theme_classic() +
    theme(
      plot.background = element_blank(),
      plot.title = black_bold_16,
      panel.background = element_rect(
        fill = "transparent",
        color = "black",
        size = 1
      ),
      axis.title.x = black_bold_16,
      axis.title.y = black_bold_16,
      axis.text.x = black_bold_16,
      axis.text.y = black_bold_16
    )
  print(eig)

  ggplot2::ggsave(
    path = file.path(output.directory, "PCA", folder),
    filename = paste("eig", x, ".pdf"),
    plot = eig,
    width = 8,
    height = 8,
    useDingbats = FALSE
  )

  corrtitle <- function(x) {
    if (x == "All") {
      title <- "PC (Healthy Tissues + Tumors)"
    } else {
      title <- paste0("PC (", x, ")")
    }
    return(title)
  }

  title <- corrtitle(x)

  var <- get_pca_var(res.pca)
  pdf(file.path(
    path = file.path(output.directory, "PCA", folder),
    filename = paste("matrix", x, ".pdf")
  ),
  width = 9,
  height = 9,
  useDingbats = FALSE
  )
  corrplot::corrplot(var$cos2, # cos2 is better than contribute
    # title = paste("PCA (", x, ")"),
    method = "color",
    title = title,
    # is.corr     = FALSE,
    tl.cex = 1.5,
    number.cex = 1.5,
    addgrid.col = "gray",
    addCoef.col = "black",
    tl.col = "black"
  )
  dev.off()
  corrplot(var$cos2, # cos2 is better than contribute
    # title = paste("PCA (", x, ")"),
    title = title,
    # is.corr = FALSE,
    tl.cex = 1.5,
    number.cex = 1.5,
    method = "color",
    addgrid.col = "gray",
    addCoef.col = "black",
    tl.col = "black"
  )


  contribplot <- function(x) {
    fviz_contrib(res.pca,
      choice = "var",
      axes = x,
      top = 10,
      fill = "lightblue",
      color = "black"
    ) +
      theme_minimal() +
      theme(
        plot.background = element_blank(),
        plot.title = black_bold_16,
        panel.background = element_rect(
          fill = "transparent",
          color = "black",
          size = 1
        ),
        axis.title.x = element_blank(),
        axis.title.y = black_bold_16,
        axis.text.x = black_bold_16_45,
        axis.text.y = black_bold_16
      )
  }
  lapply(c(1, 2), contribplot)
}

#' Plot subgroups of PCA results as biplots
#' @description This function draw biplot for PCA results
#' from TCGA and GTEX combined data from
#' \code{.get_df_subset(.TCGA_GTEX_sampletype_subset, "All")}.
#'
#' @param res.pca PCA results from from TCGA and GTEX combined RNAseq data
#' @param df combined RNAseq data with sample type annotation
#' @param x sample types for plot labeling
#' @param y type of primary diseases or healthy tissues for color of individuals
#' @param color color scheme used for individual sample on the PCA
#' @param folder sub directory name to store the output files
#' @return PCA bioplot
#' @importFrom factoextra fviz_contrib fviz_eig fviz_pca_biplot get_pca_var
#' @examples \dontrun{
#' .selected_biplot(
#'   res.pca = .RNAseq_PCA(df[[1]], 10), df = df[[2]],
#'   x = "Healthy Tissue (GTEx)", y = "sample.type", color = "#D55E00"
#' )
#' .selected_biplot(
#'   res.pca = .RNAseq_PCA(df[[1]], 10), df = df[[2]],
#'   x = "Healthy Tissue (GTEx)", y = "primary.site", color = col_vector
#' )
#' }
#' @keywords internal
#'
.selected_biplot <- function(res.pca, df, x, y, color) {
  .selected_samples <- df %>%
    dplyr::filter(.data$sample.type == x)
  biplot <- fviz_pca_biplot(res.pca,
    axes = c(1, 2),
    labelsize = 5,
    # col.ind = TCGA.GTEX.sampletype.subset$sample.type,
    col.ind = df[, y],
    palette = color,
    select.ind = list(name = row.names(.selected_samples)),
    pointshape = 20,
    pointsize = 0.75,
    # addEllipses = TRUE,
    title = "PCA - Biplot (Healthy Tissues + Tumors)",
    label = "var",
    col.var = "black",
    repel = TRUE
  ) +
    xlim(-7, 8) + ylim(-6, 7.5) + # for EIF 8
    # xlim(-6, 6) + ylim (-7, 7)+ # for EIF 4
    theme_classic() +
    theme(
      plot.background = element_blank(),
      plot.title = black_bold_16,
      panel.background = element_rect(
        fill = "transparent",
        color = "black",
        size = 1
      ),
      axis.title.x = black_bold_16,
      axis.title.y = black_bold_16,
      axis.text.x = black_bold_16,
      axis.text.y = black_bold_16,
      legend.title = element_blank(),
      legend.position = c(0, 0),
      legend.justification = c(0, 0),
      legend.background = element_blank(),
      legend.text = black_bold_16
    )
  print(biplot)
  ggplot2::ggsave(
    path = file.path(output.directory, "PCA", "All"),
    filename = paste0("EIFPCAall", x, y, ".pdf"),
    plot = biplot,
    width = 8,
    height = 8,
    useDingbats = FALSE
  )
}

## Master functions to call PCA and plot results ===============================

#' Perform PCA on RNAseq data from all tumors and healthy tissues
#' @description This function performs standard PCA on selected RNAseq data from
#' \code{EIF.list} genes generated by \code{\link{.get_df_subset}}.
#'
#' It perform PCA on tumors from all TCGA cancer types,
#' PCA on healthy tissues from all GTEx healthy tissue types,
#' and PCA on combined TCGA tumors and GTEx healthy tissues.
#'
#' This function should not be used directly, only inside \code{\link{EIF4F_PCA}} function.
#' @param EIF.list gene names in a vector of characters
#' @return PCA biplot, matrix, scree plot
#' @importFrom FactoMineR PCA
#' @keywords internal
#' @examples \dontrun{
#' .plot_PCA_TCGA_GTEX(c(
#'   "EIF4E", "EIF4G1", "EIF4A1",
#'   "EIF4EBP1", "PABPC1", "MKNK1", "MKNK2"
#' ))
#' }
.plot_PCA_TCGA_GTEX <- function(EIF.list) {
  .TCGA_GTEX_sampletype_subset <- TCGA_GTEX_RNAseq_sampletype %>%
    dplyr::select(
      all_of(EIF.list),
      "sample.type",
      "primary.disease",
      "primary.site",
      "study"
    ) %>%
    # as.tibble(.) %>%
    dplyr::filter(.data$EIF4E != 0 &
      .data$study %in% c("TCGA", "GTEX") &
      .data$sample.type %in% c(
        "Metastatic",
        "Primary Tumor",
        "Normal Tissue",
        "Solid Tissue Normal"
      )) %>%
    mutate_if(is.character, as.factor) %>%
    mutate(sample.type = factor(.data$sample.type,
      levels = c(
        "Normal Tissue",
        "Primary Tumor",
        "Metastatic",
        "Solid Tissue Normal"
      ),
      labels = c(
        "Healthy Tissue (GTEx)",
        "Primary Tumor (TCGA)",
        "Metastatic Tumor (TCGA)",
        "Adjacent Normal Tissue (TCGA)"
      )
    ))

  df <- .get_df_subset(.TCGA_GTEX_sampletype_subset, "Primary Tumor (TCGA)")
  .biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "Primary Tumor (TCGA)",
    y = "primary.disease",
    color = col_vector,
    folder = "TCGA"
  )

  df <- .get_df_subset(.TCGA_GTEX_sampletype_subset, "Metastatic Tumor (TCGA)")
  .biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "Metastatic Tumor (TCGA)",
    y = "primary.disease",
    color = col_vector,
    folder = "TCGA"
  )

  df <- .get_df_subset(.TCGA_GTEX_sampletype_subset, "Healthy Tissue (GTEx)")
  .biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "Healthy Tissue (GTEx)",
    y = "primary.site",
    color = col_vector,
    folder = "GTEX"
  )


  ## TCGA and GTEx combined
  df <- .get_df_subset(.TCGA_GTEX_sampletype_subset, "All")
  .biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "All",
    y = "sample.type",
    color = c("#D55E00", "#009E73", "#CC79A7", "#0072B2"),
    folder = "All"
  )

  .selected_biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "Healthy Tissue (GTEx)",
    y = "sample.type",
    color = "#D55E00"
  )
  .selected_biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "Healthy Tissue (GTEx)",
    y = "primary.site",
    color = col_vector
  )

  .selected_biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "Primary Tumor (TCGA)",
    y = "sample.type",
    color = "#009E73"
  )
  .selected_biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "Primary Tumor (TCGA)",
    y = "primary.disease",
    color = col_vector
  )

  .selected_biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "Metastatic Tumor (TCGA)",
    y = "sample.type",
    color = "#CC79A7"
  )
  .selected_biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "Metastatic Tumor (TCGA)",
    y = "primary.disease",
    color = col_vector
  )
}

#' Perform PCA on RNAseq data from one tumor type and matched healthy tissue
#' @description This function performs standard PCA on selected RNAseq data from
#' \code{EIF.list} genes generated by \code{\link{.get_df_subset}}.
#'
#' It perform PCA on combined tumors from TCGA cancer types and GTEx healthy tissues with the same \code{tissue} of origin.
#'
#' This function should not be used directly, only inside \code{\link{EIF4F_PCA}} function.
#' @param EIF.list gene names in a vector of characters
#' @param tissue tissue type
#' @return PCA bioplot, matrix, scree plot
#' @importFrom FactoMineR PCA
#' @keywords internal
#' @examples \dontrun{
#' .plot_PCA_TCGA_GTEX_tumor(c(
#'   "EIF4G1", "EIF4A1", "EIF4E",
#'   "EIF4EBP1", "PABPC1", "MKNK1", "MKNK2"
#' ), "Lung")
#' }
#'
.plot_PCA_TCGA_GTEX_tumor <- function(EIF.list, tissue) {
  .TCGA_GTEX_sampletype_subset <- TCGA_GTEX_RNAseq_sampletype %>%
    dplyr::select(
      all_of(EIF.list),
      "sample.type",
      "primary.disease",
      "primary.site",
      "study"
    ) %>%
    # as.data.frame(.) %>%
    dplyr::filter(.data$EIF4E != 0 &
      .data$study %in% c("TCGA", "GTEX") &
      .data$sample.type %in% c(
        "Metastatic",
        "Primary Tumor",
        "Normal Tissue",
        "Solid Tissue Normal"
      )) %>%
    mutate_if(is.character, as.factor) %>%
    mutate(sample.type = factor(.data$sample.type,
      levels = c(
        "Normal Tissue",
        "Primary Tumor",
        "Metastatic",
        "Solid Tissue Normal"
      ),
      labels = c(
        "Healthy Tissue (GTEx)",
        "Primary Tumor (TCGA)",
        "Metastatic Tumor (TCGA)",
        "Adjacent Normal Tissue (TCGA)"
      )
    )) %>%
    dplyr::filter(.data$primary.site == tissue)

  df <- .get_df_subset(.TCGA_GTEX_sampletype_subset, "All")
  .biplot(
    res.pca = .RNAseq_PCA(df[[1]], 10),
    df = df[[2]],
    x = "All",
    y = "sample.type",
    color = c("#D55E00", "#009E73", "#CC79A7", "#0072B2"),
    folder = "Lung"
  )
}

#' Perform PCA on proteomics data from CPATC LUADs and matched healthy lung tissues
#' @description This function performs standard PCA on selected proteomics data from
#' \code{CPTAC_LUAD_Proteomics} prepared by \code{\link{initialize_proteomics_data}}.
#'
#' It perform PCA on combined tumors from CPTAC LUAD and and matched healthy lung tissues.
#'
#' This function should not be used directly, only inside \code{\link{EIF4F_PCA}} function.
#' @param EIF.list gene names in a vector of characters
#' @return PCA bioplot, matrix, scree plot
#' @keywords internal
#' @examples \dontrun{
#' .plot_PCA_TCGA_GTEX_tumor(c(
#'   "EIF4G1", "EIF4A1", "EIF4E",
#'   "EIF4EBP1", "PABPC1", "MKNK1", "MKNK2"
#' ), "Lung")
#' }
#'
.plot_PCA_CPTAC_LUAD <- function(EIF.list) {
  CPTAC.LUAD.Proteomics.Sample.subset <- CPTAC_LUAD_Proteomics %>%
    column_to_rownames(var = "Sample") %>%
    mutate(Type = factor(.data$Type,
      levels = c(
        "NAT",
        "Tumor"
      ),
      labels = c(
        "Adjacent Normal Tissue (CPTAC)",
        "Primary Tumor (CPTAC)"
      )
    )) %>%
    dplyr::select(
      all_of(EIF.list),
      "Type"
    ) %>%
    # as.data.frame(.) %>%
    mutate_if(is.character, as.factor) %>%
    dplyr::filter(!is.na(.data$Type)) %>%
    tibble::remove_rownames()

  res.pca <- .protein_imputePCA(CPTAC.LUAD.Proteomics.Sample.subset, 10)

  .biplot(
    res.pca = res.pca,
    df = CPTAC.LUAD.Proteomics.Sample.subset,
    x = "LUAD(CPTAC)",
    y = "Type",
    color = c("#D55E00", "#009E73"),
    folder = "Lung"
  )
}


## wrapper function to call all master functions with inputs ===================

#' Perform PCA and generate plots
#' @description A wrapper function to call all master functions for PCA with inputs
#'
#' @details  This function run three master functions together:
#' \code{\link{.plot_PCA_TCGA_GTEX}},
#' \code{\link{.plot_PCA_TCGA_GTEX_tumor}} and
#' \code{\link{.plot_PCA_CPTAC_LUAD}} with inputs
#'
#' @return CNV analysis plots
#'
#' @export
#'
#' @examples \dontrun{
#' EIF4F_PCA()
#' }
EIF4F_PCA <- function() {
  .plot_PCA_TCGA_GTEX(c(
    "EIF4E", "EIF4G1", "EIF4A1", "EIF4EBP1",
    "PABPC1", "MKNK1", "MKNK2"
  ))

  .plot_PCA_TCGA_GTEX_tumor(
    c(
      "EIF4G1", "EIF4A1", "EIF4E", "EIF4EBP1",
      "PABPC1", "MKNK1", "MKNK2"
    ),
    "Lung"
  )

  .plot_PCA_CPTAC_LUAD(c(
    "EIF4E", "EIF4G1", "EIF4A1", "PABPC1",
    "MKNK1", "MKNK2", "EIF4EBP1"
  ))
}
